OpenRiaServices.DomainServices.Client/Framework/Silverlight/Data/IndexedList.cs

```diff
@@ -0,0 +1,158 @@
+                _dict[value] = index;
+            }
+        }
+
+        /// <summary>
+        /// Get an enumerator for items in the list.
+        /// </summary>
+        public IEnumerator GetEnumerator()
+        {
+            return _list.GetEnumerator();
+        }
+
+        /// <summary>
+        /// Copy entire list contents to the given array, starting at the
+        /// given index in the array.
+        /// </summary>
+        /// <param name="array">Array to copy this list to</param>
+        /// <param name="index">Index of given array to start copying</param>
+        public void CopyTo(Array array, int index)
+        {
+            ((IList)_list).CopyTo(array, index);
+        }
+
+        /// <summary>
+        /// Add an item to the list.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">If the passed item is null</exception>
+        /// <exception cref="InvalidOperationException">If the passed item is already in the list</exception>
+        public int Add(object value)
+        {
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+
+            if (_dict.ContainsKey(value))
+                throw new InvalidOperationException("Cannot add duplicate values");
+
+            var idx = _list.Count;
+            _list.Add(value);
+            _dict[value] = idx;
+            return idx;
+        }
+
+        /// <summary>
+        /// True if the list contains the given item.
+        /// </summary>
+        public bool Contains(object value)
+        {
+            if (value == null) throw new ArgumentNullException(nameof(value));
+            return _dict.ContainsKey(value);
+        }
+
+        /// <summary>
+        /// Remove all items from the list.
+        /// </summary>
+        public void Clear()
+        {
+            _list.Clear();
+            _dict.Clear();
+        }
+
+        /// <summary>
+        /// Get the index of the given item in the list.
+        /// </summary>
+        /// <returns>List index, or -1 if item is not in the list.</returns>
+        /// <exception cref="ArgumentNullException">If the passed item is null</exception>
+        public int IndexOf(object value)
+        {
+            if (value == null) throw new ArgumentNullException(nameof(value));
+            int idx;
+            if (_dict.TryGetValue(value, out idx))
+                return idx;
+            return NotContainedIndex;
+        }
+
+        public void Insert(int index, object value)
+        {
+            throw new NotSupportedException(
+                string.Format(CultureInfo.CurrentCulture, Resource.IsNotSupported, "Insert"));
+        }
+
+        /// <summary>
+        /// Remove the given item from the list. Does nothing if the item is
+        /// not in the list.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">If the given item is null</exception>
+        public void Remove(object value)
```

Comment:
The new implementation looks good.

If you like feel free to add an additional method "SwapRemove" or "RemoveFast" with your original code, it was a good idea and I believe we could use it in a few specific scenarios (such as no one is listening on the INotifyCollectionChanged interface).

Reply:
Also I checked out the EntitySet hierarchy and the current CreateList method could be changed to return something else such as IIndexedList/IIndexedList  (which contains the SwapRemove/FastRemove method) if it makes your work easier without it beeing a breaking change.